File: app/src/main/java/com/stathis/elmepaunivapp/ui/syllabus_lessons/recycler/LessonsAdapter.java
package com.stathis.elmepaunivapp.ui.syllabus_lessons.recycler;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.DiffItemCallbackClass;
import com.stathis.elmepaunivapp.listeners.LessonClickListener;
import com.stathis.elmepaunivapp.ui.syllabus_lessons.model.Lesson;
import com.stathis.elmepaunivapp.ui.syllabus_lessons.model.LessonHeader;

public class LessonsAdapter extends ListAdapter<Object,LessonsViewHolder> {

    public LessonsAdapter() {
        super(new DiffItemCallbackClass<Object>());
    }

    @NonNull
    @Override
    public LessonsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        return new LessonsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull LessonsViewHolder holder, int position) {
        holder.present(getItem(position));
    }

    @Override
    public int getItemViewType(int position) {
        if(getItem(position) instanceof Lesson){
            return R.layout.lesson_item_row;
        } else if (getItem(position) instanceof LessonHeader) {
            return R.layout.holder_lessons_header_item;
        } else {
            return 1;
        }
    }
}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/syllabus_lessons/recycler/LessonsAdapter.kt
package com.stathis.elmepaunivapp.ui.syllabus_lessons.recycler

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.DiffItemClass
import com.stathis.elmepaunivapp.callbacks.ElmepaClickListener
import com.stathis.elmepaunivapp.abstraction.LocalModel
import com.stathis.elmepaunivapp.databinding.HolderEmptyItemRowBinding
import com.stathis.elmepaunivapp.databinding.HolderLessonsHeaderItemBinding
import com.stathis.elmepaunivapp.databinding.LessonItemRowBinding
import com.stathis.elmepaunivapp.ui.syllabus_lessons.model.LessonHeader
import com.stathis.elmepaunivapp.ui.syllabus_lessons.model.Lesson

class LessonsAdapter(private val callback: ElmepaClickListener) : ListAdapter<LocalModel, LessonsViewHolder>(DiffItemClass<LocalModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LessonsViewHolder = when(viewType){
        R.layout.holder_lessons_header_item -> {
            val view = HolderLessonsHeaderItemBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            LessonsViewHolder(view,callback)
        }
        R.layout.lesson_item_row -> {
            val view = LessonItemRowBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            LessonsViewHolder(view,callback)
        }
        else -> {
            val view = HolderEmptyItemRowBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            LessonsViewHolder(view,callback)
        }
    }

    override fun onBindViewHolder(holder: LessonsViewHolder, position: Int) {
        holder.bindData(getItem(position))
    }

    override fun getItemViewType(position: Int): Int = when (getItem(position)) {
        is LessonHeader -> R.layout.holder_lessons_header_item
        is Lesson -> R.layout.lesson_item_row
        else -> R.layout.holder_empty_item_row
    }
}